ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -90,4 +105,153 @@ updateLoEFragStall k getCandidates =
   UpdateLoEFrag $ \ curChain _ setLoEFrag ->
     atomically $ do
       candidates <- getCandidates
-      setLoEFrag (sharedCandidatePrefix k curChain candidates)
+      setLoEFrag (fst (sharedCandidatePrefix k curChain candidates))
+
+showPeers ::
+     Show peer
+  => Map peer String
+  -> String
+showPeers = intercalate ", " . fmap (\ (peer, v) -> show peer ++ " -> " ++ v) . Map.toList
+
+densityDisconnect ::
+     ( Ord peer
+     , Show peer
+     , LedgerSupportsProtocol blk
+     )
+  => TopLevelConfig blk
+  -> ExtLedgerState blk
+  -> Map peer (AnchoredFragment (Header blk))
+  -> Map peer (Tip blk)
+  -> AnchoredFragment (Header blk)
+  -> ([peer], String)
+densityDisconnect cfg immutableLedgerSt candidateSuffixes theirTips loeFrag =
+  (losingPeers, showPeers (showBounds <$> densityBounds))
+  where
+    densityBounds = Map.fromList $ do
+      (peer, frag) <- Map.toList competingFrags
+      theirTip <- toList (theirTips Map.!? peer)
+      let candidateSuffix = candidateSuffixes Map.! peer
+          lowerBound = fromIntegral $ AF.length frag
+          unresolvedSlotsLB =
+            succWithOrigin $ AF.headSlot frag
+          unresolvedSlotsUB =
+              endOfGenesisWindow `min` claimedTip
+            where
+              claimedTip = succWithOrigin $ getTipSlotNo theirTip
+          hasBlockAfterGenesisWindow =
+            -- Note that if
+            -- > NotOrigin s = AF.headSlot candidateSuffix
+            -- this check is equivalent to
+            -- > s >= endOfGenesisWindow
+              succWithOrigin (AF.headSlot candidateSuffix)
+            > endOfGenesisWindow
```

Comment:
For blocks with a finite forecasting range (like `CardanoBlock`, but unlike `TestBlock` ATM), it can be the case that a peer already sent us a header beyond the Genesis window (equal to the forecasting range), but we don't know about this here as we only update the candidate fragment after we validated the header.

Therefore, we have to enrich the ChainSync client to also indicate whether it received, but could not yet process a header beyond the forecast horizon (ie whether it is "LedgerView-starved").

(cc @nbacquey @Niols as we just talked about this)

Reply:
Is what I just pushed coming close to what we need?

Reply:
As suggested by Nick today, it seems much more direct/easier to justify to just store the last `SlotNo` that was received in some `TVar` (either the `SlotNo` by the next header, or the rollback point), and use that here to judge whether it is beyond the Genesis window (instead of the current `AF.headSlot candidateSuffix`).

Reply:
Done in #984.